src/main/java/com/google/sps/servlets/ShoppingCartServlet.java

```diff
@@ -0,0 +1,160 @@
+    if (entity != null) {
+      // User has items added to cart.
+      books.addAll((ArrayList<String>) entity.getProperty("books"));
+    }
+    Gson gson = new Gson();
+    String json = gson.toJson(books);
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  /**
+   * Adds a new book to the current cart of the user.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get current user.
+    String uid = null;
+    JSONObject jsonObject;
+    try {
+      jsonObject = new JSONObject(request.getReader().lines().collect(Collectors.joining(System.lineSeparator())));
+    } catch (JSONException e){
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, e.getMessage());
+      return;
+    }
+    String bookId = jsonObject.get("bookId").toString();
+    try {
+      // get the idToken from hidden input field
+      uid = firebaseAuth.getUserIdFromIdToken(jsonObject.get("idToken").toString());
+    } catch (FirebaseAuthException e) {
+      response.sendError(HttpServletResponse.SC_UNAUTHORIZED, e.getMessage());
+      return;
+    }
+
+    Entity cartEntity;
+    Query query = new Query("UserCart").setFilter(new Query.FilterPredicate("uid", Query.FilterOperator.EQUAL, uid));
+    PreparedQuery results = datastore.prepare(query);
+    cartEntity = results.asSingleEntity();
+    ArrayList<String> books = new ArrayList<>();
+    if (cartEntity == null) {
+      // User cart is empty.
+      cartEntity = new Entity("UserCart");
+      cartEntity.setProperty("uid", uid);
+    } else {
+      // Add previous books to this array.
+      books.addAll((ArrayList<String>) cartEntity.getProperty("books"));
+    }
+
+    books.add(bookId);
+    cartEntity.setProperty("books", books);
+    datastore.put(cartEntity);
+    response.sendRedirect("/bookDetails.html?id=" + bookId);
+  }
+
+  /**
+   * Removes a book from the current cart of the user.
+   */
+
+  @Override
+  public void doDelete(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get current user.
+    String uid = null;
+    try {
+      // get the idToken from query string
+      uid = firebaseAuth.getUserIdFromIdToken(request.getParameter("idToken"));
+    } catch (FirebaseAuthException e) {
+      response.sendError(HttpServletResponse.SC_UNAUTHORIZED, e.getMessage());
+      return;
+    }
+    
+    Query query = new Query("UserCart").setFilter(new Query.FilterPredicate("uid", Query.FilterOperator.EQUAL, uid));
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    String bookId = request.getParameter("bookId");
+    ArrayList<String> books = new ArrayList<>();
+    for (String id : (ArrayList<String>) entity.getProperty("books")) {
+      // This ID should be deleted, so it's not added.
+      if (bookId.equals(id)) {
+        continue;
+      }
+      books.add(id);
+    }
```

Comment:
You can also use the remove function: to remove the book with id `bookId` from `books`, you can use
`books.remove(bookId);`

Here is an example: https://www.geeksforgeeks.org/remove-element-arraylist-java/

Reply:
Done, thank you for the suggestion.